ci/ray_ci/doc/autodoc.py

```diff
@@ -55,11 +56,28 @@ def _get_autodoc_rsts(self) -> List[str]:
         if self._autodoc_rsts is not None:
             return self._autodoc_rsts
 
-        dir = os.path.dirname(self._head_rst_file)
-        self._autodoc_rsts = [self._head_rst_file]
-        with open(self._head_rst_file, "r") as f:
+        self._autodoc_rsts = {self._head_rst_file}
+        visited = set()
+        while len(self._autodoc_rsts) > len(visited):
+            not_visited = self._autodoc_rsts - visited
```

Comment:
maybe just create a new node set so that you do not need to do this O(N) subtraction for the depth of d, which is O(dN)?

````
ret = {head}
current = {head}
next = set()
while current:
    for node in current:
        for c in node.children():
            if c not in ret:
               ret.add(c)
               next.add(c)
    current, next = next, set()
````

this is O(N)

(well... assuming set checking is O(1), but strictly mathematically speaking it is more like O(N x logN) , because a hash map or a bit map is not really O(1) )